submariner/IPV6-datapath.md

```diff
@@ -0,0 +1,129 @@
+
+IPv4/IPv6 dual-stack on your Kubernetes cluster provides the following features:
+
+* Dual-stack Pod networking (a single IPv4 and IPv6 address assignment per Pod)  
+* IPv4 and IPv6 enabled Services  
+* Pod off-cluster egress routing (eg. the Internet) via both IPv4 and IPv6 interfaces
+
+Currently, Submariner only supports IPV4 datapath and this proposal explains the changes required to support IPV6 and dual-stack clusters.
+
+## Inter-cluster datapath requirements
+
+The next table describes the required inter-cluster connectivity for the various combinations of clusters networking configurations:  
+
+| clusterA networking | clusterB networking | Supported connectivity type |
+| :---- | :---- | :---- |
+| V4 | V4 | V4 |
+| V4 | V6 | N/A |
+| V6 | dual-stack | V6 |
+| V6 | V6 | V6 |
+| V4 | dual-stack | V4 |
+| dual-stack | dual-stack | V4,V6 |
+
+connectivity across the clusters should be supported for the following cases:
+
+* Pods(including host-networking) to Services.
+* Pods (including host-networking) to Pods.
+
+## Proposal
+
+Currently, Submariner fully supports IPV4 inter-cluster connectivity, including egress in-cluster routing to reach GW node, GlobalNet, and inter-cluster tunnels.
+
+The idea is to duplicate intra-cluster and inter-cluster connectivity components also for IPV6.
+
+The active Gateway Engine communicates with the central Broker to advertise its Endpoint and Cluster resources to the other clusters connected to the Broker, also ensuring that it is the sole Endpoint for its cluster.
+The Endpoint resource fields should include IP addresses according to the cluster’s networking configuration.
+For example for a dual-stack cluster HealthCheckIP, PrivateIP, PublicIP and Subnets should consist of both IPv4 and IPV6 addresses.  
+
+The Route Agent running in the cluster learns about the local Endpoint and remote Endpoints and setup the necessary V4,V6 infrastructure to route cross-cluster traffic from all nodes to the active Gateway Engine node. 
+
+The active Gateway Engine also establishes a watch on the Broker to learn about the active Endpoint and Cluster resources advertised by the other clusters. Once two clusters are aware of each other’s Endpoints, they can establish a secure tunnel/s based on the content of the remote Endpoint and local Endpoint through which traffic can be routed. A tunnel should be created only if the local Endpoint networking type matches remote Endpoint ipfamily.
+
+The next diagram illustrates Submariner’s datapath architecture for kube-proxy based CNIs:
+![non-ovnk-architecture](./images/dual-stack-arch-for-non-ovnk.png)
+
+With the proposed architecture, Submariner needs to establish both V4 and V6 intra-cluster egress routing to the GW node in the case of dual-stack.
+
+Pod IPV4 egress packets for CNI != OVNK and cable-driver=libreswan will be:
+![non-ovnk-ipv4-egress](./images/ipv4-non-ovnk-egress-packets.png)
+
+And Pod IPV6 egress packets for the same configuration will be:
```

Comment:
For IPv4 vxlan tunnels, we use 242.x.x.x cidr rangle, what is the IPv6 CIDR range you are planning to use for setting up tunnels?

Reply:
I haven't checked the subject in depth, but I assume we'll find a similar address space for V6 as well.
Does it make sense?